Refactor watch functionality

CyberDem0n opened the pull request:
The old implementation has had several drawbacks:
 * It was not possible to cancel watch until the first event arrived.
 * It was creating a new watch channel for every watch.

So I've tried to fix that:
 * We create communication channels only once and all next watch and cancel requests should go via these channels.
 * it is possible to cancel existing watches without destroying the whole channel
 * depending on the method you will receive immediately either reference to a `cancel` function or a `watch_id` which might be used to cancel watch request (on the server)
 * callbacks and queues are the main building blocks now

kragniz commented on etcd3/events.py:
```diff
@@ -1,9 +1,14 @@
-        self.value = value
-        self.version = version
-        self.type = 0
+    __impl = {}
```
What's this for? It doesn't seem to get used anywhere

CyberDem0n commented:
I've updated pull request:
1. removed unused attribute `__impl`
2. did some clean up in `watch.py` and added a new method there: `stop`

`stop` method must be called when we need to stop `Wather` thread and `grpc thread` consuming watch requests stream. Currently it is called on exception in a `run` method.
